File: web/lib/scraper/renderer.ts
// lib/scraper/renderer.ts — Playwright headless como fallback para SPAs.
//
// Se usa SOLO cuando el HTML plano devuelve un shell vacío (típico en Vue/React/
// Nuxt/Next SPA sin SSR real). Renderiza con Chromium, espera networkidle,
// scrollea para disparar lazy-load, y devuelve el HTML final.

// Renderizado SPA en 3 modos:
// 1. En dev: Playwright local si está instalado (devDependency).
// 2. En prod: proxy al runtime Python (Railway) vía /render, que sí tiene Chromium.
// 3. Fallback: null → scraper degrada a fetch plano.

type RouteLike = {
  request: () => { resourceType: () => string };
  abort: () => Promise<void>;
  continue: () => Promise<void>;
};
type PageLike = {
  goto: (url: string, opts?: { waitUntil?: string; timeout?: number }) => Promise<unknown>;
  evaluate: <T>(fn: () => Promise<T> | T) => Promise<T>;
  waitForTimeout: (ms: number) => Promise<void>;
  content: () => Promise<string>;
  url: () => string;
};
type BrowserContextLike = {
  route: (pattern: string, handler: (route: RouteLike) => void) => Promise<void>;
  newPage: () => Promise<PageLike>;
  close: () => Promise<void>;
};
type BrowserContextOpts = {
  userAgent?: string;
  locale?: string;
  viewport?: { width: number; height: number };
  javaScriptEnabled?: boolean;
  ignoreHTTPSErrors?: boolean;
};
type BrowserLike = {
  newContext: (opts?: BrowserContextOpts) => Promise<BrowserContextLike>;
  close: () => Promise<void>;
};

const USER_AGENT =
  "Mozilla/5.0 (Macintosh; Intel Mac OS X 14_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36 OrdyChatBot/1.0";

let _browser: BrowserLike | null = null;
let _initPromise: Promise<BrowserLike | null> | null = null;
let _playwrightAvailable: boolean | null = null;

async function getBrowser(): Promise<BrowserLike | null> {
  if (_browser) return _browser;
  if (_playwrightAvailable === false) return null;
  if (_initPromise) return _initPromise;

  _initPromise = (async () => {
    try {
      const mod = await import("playwright").catch(() => null);
      if (!mod) {
        _playwrightAvailable = false;
        return null;
      }
      _playwrightAvailable = true;
      const b = await mod.chromium.launch({
        headless: true,
        args: ["--no-sandbox", "--disable-setuid-sandbox", "--disable-dev-shm-usage"],
      });
      _browser = b as unknown as BrowserLike;
      return _browser;
    } catch {
      _playwrightAvailable = false;
      return null;
    }
  })();
  return _initPromise;
}

export async function closeBrowser(): Promise<void> {
  if (_browser) {
    await _browser.close();
    _browser = null;
    _initPromise = null;
  }
}

export type RenderedResult = {
  url: string;
  html: string;
  durationMs: number;
  renderer: "chromium";
};

/**
 * Renderiza una URL con Chromium headless. Espera networkidle + scroll completo.
 */
export async function renderPage(url: string, timeoutMs = 25_000): Promise<RenderedResult | null> {
  // En prod: delega al runtime Python que tiene Chromium en su Docker.
  const runtimeUrl = process.env.RUNTIME_URL;
  const internalSecret = process.env.RUNTIME_INTERNAL_SECRET;
  if (runtimeUrl && internalSecret && process.env.NODE_ENV === "production") {
    try {
      const r = await fetch(`${runtimeUrl}/render`, {
        method: "POST",
        headers: { "Content-Type": "application/json", "x-internal-secret": internalSecret },
        body: JSON.stringify({ url, timeoutMs }),
        signal: AbortSignal.timeout(timeoutMs + 10_000),
      });
      if (r.ok) {
        const data = (await r.json()) as { url: string; html: string; durationMs: number };
        return { ...data, renderer: "chromium" };
      }
    } catch {
      // silent fallback
    }
  }

  const t0 = Date.now();
  const browser = await getBrowser();
  if (!browser) return null;
  const ctx = await browser.newContext({
    userAgent: USER_AGENT,
    locale: "es-ES",
    viewport: { width: 1440, height: 900 },
    javaScriptEnabled: true,
    ignoreHTTPSErrors: false,
  });

  await ctx.route("**/*", (route) => {
    const type = route.request().resourceType();
    if (["image", "media", "font", "stylesheet"].includes(type)) {
      void route.abort();
      return;
    }
    void route.continue();
  });

  const page = await ctx.newPage();
  try {
    await page.goto(url, { waitUntil: "networkidle", timeout: timeoutMs });
    await page.evaluate(async () => {
      await new Promise<void>((resolve) => {
        let total = 0;
        const step = 600;
        const id = setInterval(() => {
          window.scrollBy(0, step);
          total += step;
          if (total >= document.body.scrollHeight) {
            clearInterval(id);
            resolve();
          }
        }, 200);
      });
    });
    await page.waitForTimeout(1500);
    const html = await page.content();
    return { url: page.url(), html, durationMs: Date.now() - t0, renderer: "chromium" };
  } finally {
    await ctx.close();
  }
}

/**
 * Detecta si un HTML plano es shell vacío de SPA y por tanto necesita renderer.
 * Heurística conservadora: solo dispara el flag si está muy claro.
 */
export function needsRender(html: string, visibleTextLen: number): boolean {
  if (visibleTextLen > 3000) return false;
  // Shells SPA típicos — div contenedor vacío.
  if (/<div\s+id=["'](?:app|root|__next|__nuxt|svelte)["'][^>]*>\s*<\/div>/i.test(html)) return true;
  // Build artifacts de Vite/Webpack sin contenido visible.
  if (/<script\s+type=["']module["']/i.test(html) && visibleTextLen < 1500) return true;
  // Generator meta de frameworks SPA sin contenido.
  if (/<meta[^>]+(nuxt|gatsby|vite)[^>]*>/i.test(html) && visibleTextLen < 1500) return true;
  return false;
}

File: runtime/app/renderer.py
# runtime/app/renderer.py — Playwright headless para renderizar SPAs.
#
# Usado por el web Next.js a través del endpoint /render. Devuelve el HTML
# final tras esperar networkidle + auto-scroll (trigger lazy-load).

import asyncio
import logging
from contextlib import asynccontextmanager
from typing import Any

logger = logging.getLogger("ordychat.renderer")

# Import lazy de Playwright: en entornos sin el paquete (CI, dev sin venv
# completo) el módulo se puede importar sin romper. El error solo aparece
# al llamar renderizar().

USER_AGENT = (
    "Mozilla/5.0 (Macintosh; Intel Mac OS X 14_0) AppleWebKit/537.36 "
    "(KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36 OrdyChatBot/1.0"
)

_playwright: Any | None = None
_browser: Any | None = None
_lock = asyncio.Lock()


async def _get_browser() -> Any:
    global _playwright, _browser
    async with _lock:
        if _browser and _browser.is_connected():
            return _browser
        if _playwright is None:
            from playwright.async_api import async_playwright  # import lazy
            _playwright = await async_playwright().start()
        _browser = await _playwright.chromium.launch(
            headless=True,
            args=["--no-sandbox", "--disable-setuid-sandbox", "--disable-dev-shm-usage"],
        )
        return _browser


async def cerrar_browser() -> None:
    global _playwright, _browser
    async with _lock:
        if _browser:
            try:
                await _browser.close()
            except Exception:
                pass
            _browser = None
        if _playwright:
            try:
                await _playwright.stop()
            except Exception:
                pass
            _playwright = None


@asynccontextmanager
async def _page_ctx():
    browser = await _get_browser()
    context = await browser.new_context(
        user_agent=USER_AGENT,
        locale="es-ES",
        viewport={"width": 1440, "height": 900},
    )

    # Bloquea recursos pesados irrelevantes para acelerar render.
    async def _route(route):
        rtype = route.request.resource_type
        if rtype in ("image", "media", "font", "stylesheet"):
            await route.abort()
        else:
            await route.continue_()

    await context.route("**/*", _route)
    page = await context.new_page()
    try:
        yield page
    finally:
        await context.close()


async def renderizar(url: str, timeout_ms: int = 35_000) -> dict:
    """Renderiza una URL y devuelve {url, html, durationMs}.

    Estrategia robusta SPAs + directorios:
      1. `domcontentloaded` (rápido, HTML base listo).
      2. Auto-scroll para triggers lazy-load.
      3. Intentar `networkidle` con timeout corto para capturar hydration;
         si nunca llega (beacons/analytics vivos), seguimos con lo que hay.
      4. Esperar 1.5s extra y extraer HTML.

    `networkidle` como PRIMER paso fallaba con SPAs pesados (share.google,
    Maps, directorios con widgets). Con `domcontentloaded` + scroll cubrimos
    95% de sitios reales en 10-15s.
    """
    import time
    t0 = time.monotonic()

    async with _page_ctx() as page:
        await page.goto(url, wait_until="domcontentloaded", timeout=timeout_ms)
        # Auto-scroll para lazy-load.
        await page.evaluate("""
            async () => {
              await new Promise(resolve => {
                let total = 0;
                const step = 600;
                const id = setInterval(() => {
                  window.scrollBy(0, step);
                  total += step;
                  if (total >= document.body.scrollHeight) {
                    clearInterval(id);
                    resolve();
                  }
                }, 200);
              });
            }
        """)
        # Intento best-effort de networkidle para capturar hydration;
        # si no llega (sites con beacons siempre activos), seguimos.
        try:
            await page.wait_for_load_state("networkidle", timeout=5_000)
        except Exception:
            pass
        await page.wait_for_timeout(1500)
        html = await page.content()
        final_url = page.url

    duration_ms = int((time.monotonic() - t0) * 1000)
    return {"url": final_url, "html": html, "durationMs": duration_ms}
